pyexchange/gemini.py

```diff
@@ -0,0 +1,215 @@
+                       amount=Wad.from_number(item["amount"]))
+
+  @staticmethod
+  def from_trade(pair, item):
+    return GeminiTrade(trade_id=str(item["tid"]),
+                       timestamp=int(item["timestamp"]),
+                       pair=pair,
+                       is_sell=True if item["type"].lower() == "sell" else False,
+                       price=Wad.from_number(item["price"]),
+                       amount=Wad.from_number(item["amount"]))
+
+class GeminiOrder(Order):
+  @staticmethod
+  def create(item):
+    return GeminiOrder(order_id=item["order_id"],
+                       pair=item["symbol"].upper(),
+                       is_sell=True if item["side"].lower() == "sell" else False,
+                       price=Wad.from_number(item["price"]),
+                       amount=Wad.from_number(item["executed_amount"]),
+                       timestamp=int(item["timestamp"]))
+
+
+class GeminiApi(PyexAPI):
+  """ Gemini Exchange API interface
+
+  Implemented based on https://docs.gemini.com/rest-api/?python
+  """
+
+  logger = logging.getLogger()
+  
+  def __init__(self, api_server: str, api_key: str, api_secret: str, timeout: float) -> None:
+    assert(isinstance(api_server, str))
+    assert(isinstance(api_key, str))
+    assert(isinstance(api_secret, str))
+    assert(isinstance(timeout, float))
+
+    self.api_server = api_server
+    self.api_key = api_key
+    self.api_secret = api_secret
+    self.timeout = timeout
+
+  def get_balances(self):
+    balances = self._http_authenticated("POST", "/v1/balances")
+
+    return {
+      balance['currency']: {
+        "amount": balance["amount"],
+        "availableForTrade": balance["available"],
+        "availableForWithdrawal": balance["availableForWithdrawal"]
+      }
+      for balance in balances
+    }
+
+  def get_balance(self, coin: str):
+    assert(isinstance(coin, str))
+
+    balances = self.get_balances()
+    return balances.get(coin.upper())
+    
+  def get_orders(self, pair: str) -> List[GeminiOrder]:
+    assert(isinstance(pair, str))
+    pair = self._fix_pair(pair)
+
+    orders = self._http_authenticated("POST", "/v1/orders")
+    return list(map(lambda order: GeminiOrder.create(order), orders))
+  
+  def place_order(self, pair: str, is_sell: bool, price: Wad, amount: Wad) -> str:
+    assert(isinstance(pair, str))
+    assert(isinstance(is_sell, bool))
+    assert(isinstance(price, Wad))
+    assert(isinstance(amount, Wad))
+
+    pair = self._fix_pair(pair)
+
+    params = {
+      "symbol": pair,
+      "amount": str(amount),
+      "price": str(price),
+      "side": "sell" if is_sell else "buy",
+      "type": "exchange limit"
+    }
+    placed_order = self._http_authenticated("POST", "/v1/order/new", params)
+
+    return placed_order["order_id"]
+
+  def cancel_order(self, order_id: str, pair: str) -> bool:
+    assert(isinstance(order_id, str))
+    self.logger.info(f"Cancelling order #{order_id} on pair {pair}...")
```

Comment:
Is `pair` needed here? There's already order cancellation logging in `market-maker-keeper`, and this will lead to additional code there to override `cancel_orders_with` interface.

Reply:
Removed.